txircd/modules/extra/chanhistory.py

```diff
@@ -0,0 +1,78 @@
+from twisted.plugin import IPlugin
+from txircd.config import ConfigValidationError
+from txircd.module_interface import IMode, IModuleData, Mode, ModuleData
+from txircd.utils import ModeType, now
+from zope.interface import implements
+from datetime import timedelta
+from weakref import WeakKeyDictionary
+
+class ChanHistory(ModuleData, Mode):
+	implements(IPlugin, IModuleData, IMode)
+
+	name = "ChannelHistory"
+	affectedActions = {
+		"commandextra-PRIVMSG": 1
+	}
+	history = WeakKeyDictionary()
+	modeSettings = WeakKeyDictionary()
+
+	def channelModes(self):
+		return [ ("H", ModeType.Param, self) ]
+
+	def actions(self):
+		return [ ("modeactioncheck-channel-H-commandextra-PRIVMSG", 1, self.channelHasMode),
+		         ("modechange-channel-H", 1, self.modeChanged),
+		         ("join", 1, self.replayHistory) ]
+
+	def verifyConfig(self, config):
+		if "chanhistory_maxlines" in config:
+			if not isinstance(config["chanhistory_maxlines"], int) or config["chanhistory_maxlines"] < 0:
+				raise ConfigValidationError("chanhistory_maxlines", "invalid number")
+			elif config["chanhistory_maxlines"] > 100:
+				config["chanhistory_maxlines"] = 100
+				self.ircd.logConfigValidationWarning("chanhistory_maxlines", "value is too large", 100)
+
+	def channelHasMode(self, channel, user, data):
+		if "H" in channel.modes:
+			return channel.modes["H"]
+		return None
+
+	def modeChanged(self, channel, source, adding, param, *params):
+		if adding:
+			self.modeSettings[channel] = param.split(":")
+
+	def replayHistory(self, channel, user, fromServer):
+		if "H" not in channel.modes or channel not in self.history or channel not in self.modeSettings:
+			return
+		lines = int(self.modeSettings[channel][0])
+		seconds = int(self.modeSettings[channel][1])
+		user.sendMessage("NOTICE", "*** Replaying up to {} lines of history for {}...".format(lines, channel.name, " spanning up to {} seconds".format(seconds) if seconds > 0 else ""))
+		for messageData in self.history[channel]:
+			if now() - timedelta(seconds=seconds) < messageData[0] or seconds == 0:
+				user.sendMessage("PRIVMSG", channel.name, messageData[2], prefix=messageData[1], to="")
```

Comment:
Because this represents a historical message and our server supports server-time, you should send this with the time if the client can accept it. You can check if `"server-time" in user.cache["capabilities"]` (make sure "capabilities" exists first) and then send `{ "server-time": isoTime(messageData[0]) }` as the tags (assuming `messageData[0]` is a datetime of the time, and `isoTime` is imported from `txircd.utils`) if the client supports it.